plugins/auth-backend-module-pinniped-provider/src/authenticator.test.ts

```diff
@@ -0,0 +1,503 @@
+        startRequest,
+        implementation,
+      );
+      const { searchParams } = new URL(startResponse.url);
+      const stateParam = searchParams.get('state');
+      const decodedState = decodeOAuthState(stateParam!);
+
+      expect(decodedState).toMatchObject(oauthState);
+    });
+
+    it('fails when request has no session', async () => {
+      return expect(
+        pinnipedAuthenticator.start(
+          {
+            state: encodeOAuthState(oauthState),
+            req: {
+              method: 'GET',
+              url: 'test',
+            },
+          } as unknown as OAuthAuthenticatorStartInput,
+          implementation,
+        ),
+      ).rejects.toThrow('authentication requires session support');
+    });
+  });
+
+  describe('#authenticate', () => {
+    let handlerRequest: OAuthAuthenticatorAuthenticateInput;
+
+    beforeEach(() => {
+      handlerRequest = {
+        req: {
+          method: 'GET',
+          url: `https://test?code=authorization_code&state=${encodeOAuthState(
+            oauthState,
+          )}`,
+          session: {
+            'oidc:pinniped.test': {
+              state: encodeOAuthState(oauthState),
+            },
+          },
+        } as unknown as express.Request,
+      };
+    });
+
+    it('exchanges authorization code for access token', async () => {
+      const handlerResponse = await pinnipedAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const accessToken = handlerResponse.session.accessToken;
+
+      expect(accessToken).toEqual('accessToken');
+    });
+
+    it('exchanges authorization code for refresh token', async () => {
+      const handlerResponse = await pinnipedAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const refreshToken = handlerResponse.session.refreshToken;
+
+      expect(refreshToken).toEqual('refreshToken');
+    });
+
+    it('returns granted scope', async () => {
+      const handlerResponse = await pinnipedAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+      const responseScope = handlerResponse.session.scope;
+
+      expect(responseScope).toEqual('testScope');
+    });
+
+    it('returns cluster-scoped ID token when audience is specified', async () => {
+      oauthState.audience = 'test_cluster';
+      handlerRequest = {
+        req: {
+          method: 'GET',
+          url: `https://test?code=authorization_code&state=${encodeOAuthState(
+            oauthState,
+          )}`,
+          session: {
+            'oidc:pinniped.test': {
+              state: encodeOAuthState(oauthState),
+            },
+          },
+        } as unknown as express.Request,
+      };
+
+      const handlerResponse = await pinnipedAuthenticator.authenticate(
+        handlerRequest,
+        implementation,
+      );
+
+      expect(handlerResponse.session.idToken).toEqual(clusterScopedIdToken);
+    }, 70000);
```

Comment:
oops, don't need this giant timeout
```suggestion
    });
```

Reply:
Done